xds/internal/xdsclient/authority.go

```diff
@@ -39,44 +35,60 @@ import (
-	md              xdsresource.UpdateMetadata           // Metadata for the most recent update
-	deletionIgnored bool                                 // True if resource deletion was ignored for a prior update
-
-	// Common watch state for all watchers of this resource.
-	wTimer *time.Timer // Expiry timer
-	wState watchState  // State of the watch
+	watchers          map[xdsresource.ResourceWatcher]bool // Set of watchers for this resource.
+	cache             xdsresource.ResourceData             // Most recent ACKed update for this resource.
+	md                xdsresource.UpdateMetadata           // Metadata for the most recent update.
+	deletionIgnored   bool                                 // True, if resource deletion was ignored for a prior update.
+	xdsChannelConfigs []*xdsChannelWithConfig              // List of xdsChannels where this resource is subscribed.
+}
+
+// xdsChannelForADS is used to acquire a reference to an xdsChannel. This
+// functionality is provided by the xdsClient.
+//
+// The arguments to the function are as follows:
+//   - the server config for the xdsChannel
+//   - the calling authority on which a set of callbacks are invoked by the
+//     xdsChannel on ADS stream events
+//
+// Returns a reference to the xdsChannel and a function to release the same. A
+// non-nil error is returned if the channel creation fails and the first two
+// return values are meaningless in this case.
+type xdsChannelForADS func(*bootstrap.ServerConfig, *authority) (*xdsChannel, func(), error)
+
+// xdsChannelWithConfig is a struct that holds an xdsChannel and its associated
+// ServerConfig, along with a cleanup function to release the xdsChannel.
+type xdsChannelWithConfig struct {
+	xc      *xdsChannel
+	sc      *bootstrap.ServerConfig
+	cleanup func()
 }
 
-// authority wraps all state associated with a single management server. It
-// contains the transport used to communicate with the management server and a
-// cache of resource state for resources requested from the management server.
+// authority provides the functionality required to communicate with a
+// management server corresponding to an authority name specified in the
+// bootstrap configuration.
 //
-// Bootstrap configuration could contain multiple entries in the authorities map
-// that share the same server config (server address and credentials to use). We
-// share the same authority instance amongst these entries, and the reference
-// counting is taken care of by the `clientImpl` type.
+// It holds references to one or more xdsChannels, one for each server
+// configuration in the bootstrap, to allow fallback from a primary management
+// server to a secondary management server. Authorities that contain similar
+// server configuration entries will end up sharing the xdsChannel for that
+// server configuration. The xdsChannels are owned and managed by the xdsClient.
+//
+// It also contains cache of resource state for resources requested from the
```

Comment:
Nit: s/the/a. And management server(s) should be plural since numerous servers can be present here?